### ISSUE TITLE
CudaInterOp unit tests fail when run on device id != 0

### ISSUE THREAD

**Issue by ndellingwood** [kokkos/kokkos#5632]
**Describe the bug**
In a cuda/11.0+gcc/9.2.0 build, the `KokkosCore_UnitTest_CudaInterOpInit` and `KokkosCore_UnitTest_CudaInterOpStreams` unit tests failed when setting `KOKKOS_DEVICE_ID=1` on a system with two V100 GPUs

**nvidia-smi output**:
```
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 515.43.04    Driver Version: 515.43.04    CUDA Version: 11.7     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  Tesla V100-PCIE...  Off  | 00000000:5E:00.0 Off |                    0 |
| N/A   33C    P0    35W / 250W |      0MiB / 16384MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
|   1  Tesla V100-PCIE...  Off  | 00000000:D8:00.0 Off |                    0 |
| N/A   32C    P0    37W / 250W |      0MiB / 16384MiB |      2%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
                                                                               
+-----------------------------------------------------------------------------+
| Processes:                                                                  |
|  GPU   GI   CI        PID   Type   Process name                  GPU Memory |
|        ID   ID                                                   Usage      |
|=============================================================================|
|  No running processes found                                                 |
+-----------------------------------------------------------------------------+
```

**Test failures**:
```
[ndellin@kokkos-dev-2 Cuda11Gcc920-deviceid_1]$ KOKKOS_DEVICE_ID=1 ctest -R KokkosCore_UnitTest_CudaInterOpInit --output-on-failure
Test project /ascldap/users/ndellin/kokkos/testing/Cuda11Gcc920-deviceid_1
    Start 8: KokkosCore_UnitTest_CudaInterOpInit
1/1 Test #8: KokkosCore_UnitTest_CudaInterOpInit ...Child aborted***Exception:   2.66 sec
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from cuda
[ RUN      ] cuda.raw_cuda_interop
cudaDeviceSynchronize() error( cudaErrorIllegalAddress): an illegal memory access was encountered /ascldap/users/ndellin/kokkos/core/src/Cuda/Kokkos_Cuda_Instance.cpp:161
Backtrace:
                        [0x46b223]
                        [0x4662b8]
                        [0x4662eb]
                        [0x46d9bd]
                        [0x471878]
                        [0x44ad35]
                        [0x415d0b]
                        [0x412c14]
                        [0x4493bd]
                        [0x43afca]
                        [0x43b43d]
                        [0x43bb31]
                        [0x43d20a]
                        [0x43d6eb]
                        [0x410abf]
__libc_start_main [0x7f048397b555]
                        [0x41250d]


0% tests passed, 1 tests failed out of 1

Total Test time (real) =   2.66 sec

The following tests FAILED:
	  8 - KokkosCore_UnitTest_CudaInterOpInit (Child aborted)


[ndellin@kokkos-dev-2 Cuda11Gcc920-deviceid_1]$ KOKKOS_DEVICE_ID=1 ctest -R KokkosCore_UnitTest_CudaInterOpStreams --output-on-failure
Test project /ascldap/users/ndellin/kokkos/testing/Cuda11Gcc920-deviceid_1
    Start 9: KokkosCore_UnitTest_CudaInterOpStreams
1/1 Test #9: KokkosCore_UnitTest_CudaInterOpStreams ...***Failed  Error regular expression found in output. Regex=[  FAILED  ]  2.25 sec
[==========] Running 1 test from 1 test suite.
[----------] Global test environment set-up.
[----------] 1 test from cuda
[ RUN      ] cuda.raw_cuda_streams
/ascldap/users/ndellin/kokkos/core/unit_test/cuda/TestCuda_InterOp_Streams.cpp:73: Failure
Expected equality of these values:
  600
  sum
    Which is: 0
[  FAILED  ] cuda.raw_cuda_streams (1671 ms)
[----------] 1 test from cuda (1671 ms total)

[----------] Global test environment tear-down
[==========] 1 test from 1 test suite ran. (1671 ms total)
[  PASSED  ] 0 tests.
[  FAILED  ] 1 test, listed below:
[  FAILED  ] cuda.raw_cuda_streams

 1 FAILED TEST
Kokkos::Cuda ERROR: Failed to call Kokkos::Cuda::finalize()


0% tests passed, 1 tests failed out of 1

Total Test time (real) =   2.25 sec

The following tests FAILED:
	  9 - KokkosCore_UnitTest_CudaInterOpStreams (Failed)
```

**Please also include the following items to support reproducing the bug**
**Reproducer**:
SHA e78d6a53a21da68a61cf1fd9dcefcb98fc161ad4
```bash
module load sems-archive-env sems-archive-cmake/3.17.1 sems-archive-gcc/9.2.0 cuda/11.0

$KOKKOS_PATH/generate_makefile.bash --with-devices=Cuda,Serial --arch=Volta70 --compiler=$KOKKOS_PATH/bin/nvcc_wrapper --cxxflags="-O3 -Wall -Wunused-parameter -Wshadow -pedantic -Werror -Wsign-compare -Wtype-limits -Wuninitialized " --cxxstandard="17" --kokkos-path=$KOKKOS_PATH --with-cuda-options=enable_lambda  --no-examples
```

**Comment by masterleinad** [comment]
I would expect
```diff
diff --git a/core/unit_test/cuda/TestCuda_InterOp_Init.cpp b/core/unit_test/cuda/TestCuda_InterOp_Init.cpp
index 31fd63f08..c9e496bd9 100644
--- a/core/unit_test/cuda/TestCuda_InterOp_Init.cpp
+++ b/core/unit_test/cuda/TestCuda_InterOp_Init.cpp
@@ -59,9 +59,10 @@ __global__ void offset(int* p) {
 // Test whether allocations survive Kokkos initialize/finalize if done via Raw
 // Cuda.
 TEST(cuda, raw_cuda_interop) {
+  Kokkos::initialize();
+
   int* p;
   KOKKOS_IMPL_CUDA_SAFE_CALL(cudaMalloc(&p, sizeof(int) * 100));
-  Kokkos::initialize();
 
   Kokkos::View<int*, Kokkos::MemoryTraits<Kokkos::Unmanaged>> v(p, 100);
   Kokkos::deep_copy(v, 5);
diff --git a/core/unit_test/cuda/TestCuda_InterOp_Streams.cpp b/core/unit_test/cuda/TestCuda_InterOp_Streams.cpp
index f11f657e0..432a3f0bf 100644
--- a/core/unit_test/cuda/TestCuda_InterOp_Streams.cpp
+++ b/core/unit_test/cuda/TestCuda_InterOp_Streams.cpp
@@ -48,9 +48,9 @@
 namespace Test {
 // Test Interoperability with Cuda Streams
 TEST(cuda, raw_cuda_streams) {
+  Kokkos::initialize();
   cudaStream_t stream;
   cudaStreamCreate(&stream);
-  Kokkos::initialize();
   int* p;
   cudaMalloc(&p, sizeof(int) * 100);
   using MemorySpace = typename TEST_EXECSPACE::memory_space;
```
to fix the problem. Can you confirm?

**Comment by masterleinad** [comment]
Basically, we need to set the device we want to operate on first and we do that in `Kokkos::initialize`.

**Comment by ndellingwood** [comment]
> Can you confirm?

@masterleinad yes, that resolved the failures, thanks! Are you able to put in a PR, or would you like me to and credit you as the author?

**Comment by ndellingwood** [comment]
PR #5633 up with the suggested fix

**Comment by PhilMiller** [comment]
Is the failure reproducible on 3.7? If so, this should not be closed until 4.0 is released (or until the PR gets cherry picked to a 3.7.2 patch release, I guess)

**Comment by ndellingwood** [comment]
Reproducible with 3.7.01, the fix in #5634 will need cherry-pick to 3.7.02
